tests/topsql_test.go

```diff
@@ -0,0 +1,408 @@
+package tests
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/app/vminsert"
+	"github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect"
+	"github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql"
+	"github.com/VictoriaMetrics/VictoriaMetrics/app/vmstorage"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/storage"
+	"github.com/genjidb/genji"
+	"github.com/gin-gonic/gin"
+	"github.com/golang/protobuf/proto"
+	rua "github.com/pingcap/kvproto/pkg/resource_usage_agent"
+	"github.com/pingcap/ng_monitoring/component/topology"
+	"github.com/pingcap/ng_monitoring/component/topsql"
+	"github.com/pingcap/ng_monitoring/component/topsql/query"
+	"github.com/pingcap/ng_monitoring/component/topsql/service"
+	"github.com/pingcap/ng_monitoring/component/topsql/store"
+	"github.com/pingcap/ng_monitoring/config"
+	"github.com/pingcap/ng_monitoring/config/pdvariable"
+	"github.com/pingcap/ng_monitoring/database/timeseries"
+	"github.com/pingcap/tipb/go-tipb"
+	"github.com/stretchr/testify/suite"
+)
+
+func TestTopSQL(t *testing.T) {
+	suite.Run(t, &testTopSQLSuite{})
+}
+
+// TODO: It is not elegant to open the local port for testing,
+//       we may need to refactor the code to achieve loose-coupling.
+const (
+	testTsdbPath       = "/tmp/ng-monitoring-test/tsdb"
+	testTiKVServerIP   = "127.0.0.1"
+	testTiKVServerPort = 10283
+)
+
+var (
+	testTiKVAddr = fmt.Sprintf("%s:%d", testTiKVServerIP, testTiKVServerPort)
```

Comment:
how about listening to a random port? since port `10283` maybe already used.
```
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	...
        listener.Addr().String() // get address

```

Reply:
Good idea, done